### ISSUE TITLE
MbedTLS error CTR_DRBG - The requested random buffer length is too big

### ISSUE THREAD

**Issue by M95D** [transmission/transmission#5795]
### What is the issue?

I get lots of these errors in the log:
```ERR crypto-utils-mbedtls.cc:232 MbedTLS error: CTR_DRBG - The requested random buffer length is too big (-54) (crypto-utils-mbedtls.cc:232```

transmission-daemon 4.0.3, running on OpenWrt snapshot r22977, 64bit, on a Raspberry Pi 3B+

Thank you.

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

4.0.3 (6b0e49bbb2)

**Comment by luizluca** [comment]
mbedtls limits the output buffer of random values to 1024 bytes.

And I believe this line might be called with larger values:

https://github.com/transmission/transmission/blob/d7473f22ed1f1cc637a23d17f12f3c22d5c4b8a4/libtransmission/crypto-utils-mbedtls.cc#L223-L234

If that is what is causing the issue, it should not cause harm (other than log spamming) as it will fall back to tr_rand_buffer_std.

(https://github.com/transmission/transmission/blob/d7473f22ed1f1cc637a23d17f12f3c22d5c4b8a4/libtransmission/crypto-utils.cc#L252-L258)

Anyway, transmission should break the request into smaller blocks or, at least, directly fallback without spamming the log. Something like this after the (length == 0) check:

```c
    if (length > MBEDTLS_CTR_DRBG_MAX_REQUEST)
         return false;

```

**Comment by abelian424** [comment]
I'm running transmission 4.0.3 on OpenWrt 23.0.5-rc3 with OTA software updates and transmission seems to crash after a while. I have logging set at "Info" and there's no indication of why it crashed, and this is the only change in functioning that I have noticed. It's possible that transmission stops logging all errors altogether due to the log spamming.

> ERR crypto-utils-mbedtls.cc:232 Too many messages like this! I won't log this message anymore this session.

**Comment by abelian424** [comment]
I have `ftruncate` and `copy_file_range`syscalls in /etc/seccomp but I still have this problem and it results in a crash loop making transmission unusable on an OpenWrt router with 512MB ram.

**Comment by M95D** [comment]
Build your own custom OpenWrt with audit enabled and use that to find which syscall needs to be added. See https://github.com/openwrt/packages/pull/21641 for the final part.

**Comment by luizluca** [comment]
> strace

It does give a good hint

**Comment by siiky** [comment]
I still get tons of error messages like this in the latest versions (4.0.5 and 4.0.6) on my RPi 1:

```
Jul 08 13:20:11 hostname transmission-daemon[19072]: [2024-07-08 13:20:11.518] ERR crypto-utils-mbedtls.cc:232 MbedTLS error: CTR_DRBG - The requested random buffer length is too big (-54) (crypto-utils-mbedtls.cc:232)
Jul 08 13:20:11 hostname transmission-daemon[19072]: [2024-07-08 13:20:11.518] ERR crypto-utils-mbedtls.cc:232 MbedTLS error: CTR_DRBG - The requested random buffer length is too big (-54) (crypto-utils-mbedtls.cc:232)
Jul 08 13:20:11 hostname transmission-daemon[19072]: [2024-07-08 13:20:11.518] ERR crypto-utils-mbedtls.cc:232 MbedTLS error: CTR_DRBG - The requested random buffer length is too big (-54) (crypto-utils-mbedtls.cc:232)
Jul 08 13:20:11 hostname transmission-daemon[19072]: [2024-07-08 13:20:11.518] ERR crypto-utils-mbedtls.cc:232 MbedTLS error: CTR_DRBG - The requested random buffer length is too big (-54) (crypto-utils-mbedtls.cc:232)
Jul 08 13:20:11 hostname transmission-daemon[19072]: [2024-07-08 13:20:11.518] ERR crypto-utils-mbedtls.cc:232 MbedTLS error: CTR_DRBG - The requested random buffer length is too big (-54) (crypto-utils-mbedtls.cc:232)
Jul 08 13:20:11 hostname transmission-daemon[19072]: [2024-07-08 13:20:11.518] ERR crypto-utils-mbedtls.cc:232 MbedTLS error: CTR_DRBG - The requested random buffer length is too big (-54) (crypto-utils-mbedtls.cc:232)
Jul 08 13:20:11 hostname transmission-daemon[19072]: [2024-07-08 13:20:11.518] ERR crypto-utils-mbedtls.cc:232 MbedTLS error: CTR_DRBG - The requested random buffer length is too big (-54) (crypto-utils-mbedtls.cc:232)
Jul 08 13:20:11 hostname transmission-daemon[19072]: [2024-07-08 13:20:11.518] ERR crypto-utils-mbedtls.cc:232 MbedTLS error: CTR_DRBG - The requested random buffer length is too big (-54) (crypto-utils-mbedtls.cc:232)
Jul 08 13:20:11 hostname transmission-daemon[19072]: [2024-07-08 13:20:11.519] ERR crypto-utils-mbedtls.cc:232 MbedTLS error: CTR_DRBG - The requested random buffer length is too big (-54) (crypto-utils-mbedtls.cc:232)
Jul 08 13:20:11 hostname transmission-daemon[19072]: [2024-07-08 13:20:11.519] ERR crypto-utils-mbedtls.cc:232 Too many messages like this! I won't log this message anymore this session. (crypto-utils-mbedtls.cc:232)
```

These are the config options I used to build Transmission:

```
cmake -DCMAKE_BUILD_TYPE=MinSizeRel -DENABLE_GTK=OFF -DENABLE_QT=OFF -DENABLE_MAC=OFF -DREBUILD_WEB=OFF -DINSTALL_WEB=OFF -DENABLE_TESTS=OFF -DENABLE_NLS=OFF -DINSTALL_DOC=OFF -DWITH_APPINDICATOR=OFF -DCMAKE_C_FLAGS='-Os -march=native' -DCMAKE_CXX_FLAGS='-Os -march=native' -DWITH_INOTIFY=OFF -DCMAKE_INSTALL_PREFIX=$HOME/.local ..
```

**Comment by siiky** [comment]
Apologies for the noise, I only now had the time to investigate and realized that #6379 hasn't been included in v4.0.6. I wonder how v4.0.5 was working a week or so ago...

@ckerr the fix is in the `main` branch, but not in the v4.0.6 release. Was that intended? If so, anything I should be worried about?
